File: Java/src/deepspace/NumericDamage.java
package deepspace;

import java.util.ArrayList; // import the ArrayList class

/**
 *
 * @author Mario Megías Mateo and Shao Jie Hu Chen.
 */
public class NumericDamage extends Damage {
    
    private int nWeapons;
    
    /**
     * @brief Constructor
     * @param s Number of shields
     * @param w Number of weapons
     */
    NumericDamage(int w, int s) {
        super(s);
        nWeapons = w;
    }
    
    /**
     * @brief Copies the implicit object. ABSTRACT.
     * @return A copy of the implicit object.
     */
    @Override
    public NumericDamage copy() {
        return new NumericDamage(nWeapons, getNShields());
    }
    
    /**
     * @brief UI Interface. Overrride.
     * @return Interface to UI
     */
    @Override
    NumericDamageToUI getUIversion() {
        return new NumericDamageToUI(this);
    }
    
    /**
     * @brief It reduces the Damage according to restrictions provided by w 
     * and s. Override.
     * @param w ArrayList of Weapons to adjust.
     * @param s ArrayList of ShieldsBooster to adjust.
     * @return New instance of Damage, with adjusted restrictions.
     */
    @Override
    public NumericDamage adjust(ArrayList<Weapon> w, ArrayList<ShieldBooster> s) {
        int numberofShields = adjustShields(s);
        int numberofWeapons = Math.min(w.size(), this.nWeapons);
        return new NumericDamage(numberofWeapons, numberofShields);
    }
    
    /**
     * @brief It discards a Weapon from the Damage. Override.
     * @param w
     */
    @Override
    public void discardWeapon(Weapon w) {
        if (this.nWeapons > 0)
            this.nWeapons--;
        else
            this.nWeapons = 0;
    }
        
    /**
     * @brief Establishes when a Damage is inoffensive.
     * @return True when there are not shields not weapons. False otherwise. 
     */
    @Override
    public boolean hasNoEffect() {
        return (super.hasNoEffect() && nWeapons == 0);
    }
    
    /**
     * @brief Getter
     * @return nWeapons attribute
     */
    public int getNWeapons() {
        return nWeapons;
    }
    
    /**
     * @brief toString
     * @return A string with the information of the state of the object
     */
    @Override
    public String toString() {
        String cadSuper = super.toString();
        String info = "";
        info += cadSuper.substring(0, cadSuper.length()-1);
        info += ", nWeapons: " + nWeapons;
        info += "]";
        return info;
    }
             
}

File: Ruby/NumericDamage.rb
#encoding:utf-8

# +-------------------------------------------------------------------+ #
# | DEEPSPACE (Ruby)                                                  | #
# | Authors: Shao Jie Hu Chen and Mario Megías Mateo                  | #
# +-------------------------------------------------------------------+ #
# | Numeric Damage.                                                   | #
# +-------------------------------------------------------------------+ #

module Deepspace

require_relative 'Damage.rb'
require_relative 'Weapon.rb'
require_relative 'ShieldBooster.rb'
require_relative 'NumericDamageToUI.rb'

class NumericDamage < Damage
    
    # Override
    def initialize(theWeapons, theShields)
        super(theShields)
        @nWeapons = theWeapons
    end
    public_class_method :new

    attr_reader :nWeapons

    # Override
    def copy
        return NumericDamage.new(nWeapons, nShields)
    end

    # Override
    def getUIversion
        return NumericDamageToUI.new(self)
    end

    def adjustWeapons(theWeaponsArray)
        return [nWeapons,theWeaponsArray.length].min
    end

    private :adjustWeapons

    # Override
    def adjust(theWeaponsArray, theShieldsArray)
        numShields = super
        # numShields = newDamage.nShields
        numWeapons = adjustWeapons(theWeaponsArray)

        return NumericDamage.new(numWeapons, numShields)
    end

    # Override
    def discardWeapon(weapon)
        if @nWeapons > 0 
            @nWeapons -= 1
        else
            @nWeapons = 0
        end
    end

    # Override
    def hasNoEffect
        return (super && (nWeapons == 0))
    end
    
    # Override
    def to_s
        s = super

        s2 = s.sub("]", "")
        s2 += ", nWeapons: #{nWeapons}]"

        return s2
    end

end # End NumericDamage
end # End Deepspace

if $0 == __FILE__ then

    puts "--------------------------------------"
    puts "Constructor                           "
    puts "--------------------------------------"
    
    numDamage = Deepspace::NumericDamage.new(2,1)
    puts numDamage.to_s
    puts numDamage.inspect

    puts "\n\n"

    puts "--------------------------------------"
    puts "Copy                                  "
    puts "--------------------------------------"

    copyDamage = numDamage.copy

    puts copyDamage.inspect

    puts "\n\n"

    puts "--------------------------------------"
    puts "UI Version                            "
    puts "--------------------------------------"

    uiDamage = numDamage.getUIversion

    puts uiDamage.inspect

    puts "\n\n"

    puts "--------------------------------------"
    puts "Adjust                                "
    puts "--------------------------------------"

    s1 = Deepspace::ShieldBooster.new("Shield",2,3)
    w1 = Deepspace::Weapon.new("Weapon",Deepspace::WeaponType::LASER, 5)

    puts numDamage.adjust([w1,w1,w1], [s1,s1])

    puts "\n\n"

    puts "--------------------------------------"
    puts "Discards                              "
    puts "--------------------------------------"

    numDamage.discardShieldBooster
    numDamage.discardWeapon(2)

    puts numDamage.to_s

    # We shall empty the Damage shields and weapons.

    numDamage.discardShieldBooster
    numDamage.discardWeapon(2)
    numDamage.discardShieldBooster
    numDamage.discardWeapon(2)
    numDamage.discardShieldBooster
    numDamage.discardWeapon(2)

    puts numDamage.to_s + " (should be null)"

    puts "\n\n"

    puts "--------------------------------------"
    puts "NoEffect                              "
    puts "--------------------------------------"

    if numDamage.hasNoEffect
        puts "No effect (as expected)."
    else
        puts "Effect (NOT expected)."
    end
end  